File: Java/2542. Maximum Subsequence Score.java
class Solution {
    public long maxScore(int[] nums1, int[] nums2, int k) {
        int n = nums1.length;
        int[][] pairs = new int[n][2];

        for (int i=0; i<n; i++) {
            pairs[i] = new int[] {nums1[i], nums2[i]};
        }
        Arrays.sort(pairs, (a, b) -> b[1] - a[1]);

        PriorityQueue<Integer> heap = new PriorityQueue<>(k, (a, b) -> a - b);
        long cur_sum = 0;
        for (int i=0; i<k; i++) {
            cur_sum += pairs[i][0];
            heap.add(pairs[i][0]);
        }

        long ans = cur_sum * pairs[k - 1][1];
        for (int i=k; i<n; i++) {
            cur_sum += pairs[i][0] - heap.poll();
            heap.offer(pairs[i][0]);
            ans = Math.max(ans, cur_sum * pairs[i][1]);
        }

        return ans;
    }
}

File: Python3/2542. Maximum Subsequence Score.py
class Solution:
    def maxScore(self, nums1: List[int], nums2: List[int], k: int) -> int:
        pairs = [(a, b) for a, b in zip(nums1, nums2)]
        # sort decreasingly on nums2
        pairs.sort(key = lambda x: -x[1])

        # use a min heap to maintain the k largest numbers
        heap = [x[0] for x in pairs[:k]]
        cur_sum = sum(heap)
        heapq.heapify(heap)
        ans = cur_sum * pairs[k - 1][1]

        for n1, n2 in pairs[k:]:
            # remove the smallest integer from the previous k largest numbers
            # and push the current i-th number from nums1
            cur_sum -= heapq.heappop(heap)
            cur_sum += n1
            heapq.heappush(heap, n1)
            ans = max(ans, cur_sum * n2)

        return ans